### PR TITLE
Add derivative for cholesky_inverse

### LINKED ISSUE #4669: [Feature request] Gradient of cholesky_inverse, cholesky_solve
Hello everyone, 

Thank you for your amazing work. 
I would be interested by the implementation of the first derivative of the potri or/and potrs operators. I can use matrix inversion instead, but it comes with a performance drop cost.

Thank you again, 
Best, 

Alex

### PR DESCRIPTION
**Pull request by vishwakftw**
Changelog:

- Add derivative of cholesky_inverse. The equations are derived akin to the derivative of solve methods using the technique detailed [here](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=2ahUKEwiXrOjIyM7kAhWstlkKHRxqCDgQFjAAegQIAhAC&url=https%3A%2F%2Fpeople.maths.ox.ac.uk%2Fgilesm%2Ffiles%2FNA-08-01.pdf&usg=AOvVaw0BNISOvM_I9KjPrl0xv1R_)

Test Plan:

- Added tests for cholesky_inverse in test_autograd.py

Closes #4669.

### CODE REVIEW

**Comment by vishwakftw** [comment]
cc: @Balandat @jacobrgardner @fritzo

File: test/test_autograd.py [hunk 1]
```diff
@@ -2564,6 +2564,25 @@ def run_test(upper, dims):
         for upper, dims in product([True, False], [(3, 3), (5, 3, 3), (4, 3, 2, 2)]):
             run_test(upper, dims)
 
+    @skipIfNoLapack
+    def test_cholesky_inverse(self):
+        def _test_with_size(upper, dims):
+            # Too small values cause inconsistencies in gradient calculation, hence the special initialization
```
Comment (Balandat) [inline test/test_autograd.py]: Let' be explicit that we need to construct a Cholesky factor here, not just any matrix without "too small values". Might also make sense to allow negative off-diagonal entries in the tests. The issue with the diagonal entries being close to zero is just that the matrix will be almost singular.
Something along the lines of 
```
A_ = torch.randn(*dims)
A_.diagonal().uniform_(0.1, 5)
A_.requires_grad_()
```

I'm using `A_` here because `root` is confusing (the triangular part of it is what's considered the root - i.e. `A` in `func`).

**Review by Balandat — COMMENTED** [review]
So the issue then was just with the test matrices? Apart from my other comment I think this looks good.

File: test/test_autograd.py (hunk 1 above)
Comment (vishwakftw) [reply test/test_autograd.py]: Thank you for your comment, I'll fix this.

File: test/test_autograd.py [hunk 2]
```diff
@@ -2442,6 +2442,29 @@ def run_test(upper, dims):
...
+                if upper:
+                    root = A.triu()
+                else:
+                    root = A.tril()
+                return torch.cholesky_inverse(root, upper)
+
+            gradcheck(func, [A, upper])
+            gradgradcheck(func, [A, upper])
+
+        for upper, dims in product([True, False], [(3, 3), (5, 5)]):
```
Comment (fritzo) [inline test/test_autograd.py]: Why not also check batched version with dims e.g. `(3, 2, 2)`?

**Review by fritzo — COMMENTED** [review]
The tests look good to me, but I am linalg-dyslexic and always defer to @fehiepsi for review :smile:

File: test/test_autograd.py (hunk 2 above)
Comment (vishwakftw) [reply test/test_autograd.py]: We don’t have batching for cholesky_inverse yet, working on it.

**Comment by fehiepsi** [comment]
Oh, this is nice! @vishwakftw just for my curiosity, could you point out which part of [the article](https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf) which you used to derive the gradient?

**Comment by vishwakftw** [comment]
I used 2.2.3 Matrix Inverse, where C = LL^T

**Comment by vishwakftw** [comment]
@Balandat also verified the answer

**Comment by vishwakftw** [comment]
Test failures are legitimate, I’ll investigate in a while.

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -751,6 +751,22 @@ Tensor cholesky_backward(Tensor grad, bool upper, Tensor L) {
...
 
+Tensor cholesky_inverse_backward(Tensor grad, Tensor L, bool upper, Tensor inverse) {
+  Tensor grad_L;
+  if (grad.defined()) {
+    Tensor common_term = grad + grad.transpose(-2, -1);
+    common_term = at::matmul(inverse, at::matmul(common_term, inverse));
+    if (upper) {
+      grad_L = -at::matmul(inverse, common_term);
+    } else {
+      grad_L = -at::matmul(common_term, inverse);
```
Comment (fehiepsi) [inline tools/autograd/templates/Functions.cpp]: is this `at::matmul(common_term, L);`?

Comment (fehiepsi) [inline tools/autograd/templates/Functions.cpp]: Maybe I am wrong because the tests pass, but from the reference, I think this is `-at::matmul(common_term, L)`

Comment (vishwakftw) [reply tools/autograd/templates/Functions.cpp]: Oh yes, I botched that in a renaming update. Thanks for pointing it out!

**Comment by fehiepsi** [comment]
Thansk @vishwakftw, it is the first time I know about how these matrix gradients are manipulated. :D LGTM!

**Comment by Balandat** [comment]
@fehiepsi, fwiw here is a little writeup of the derivation: [Cholesky_Inverse_gradient.pdf](https://github.com/pytorch/pytorch/files/3643896/Cholesky_Inverse_gradient.pdf)

**Comment by fehiepsi** [comment]
Thanks @Balandat !

**Review by Balandat — APPROVED** [review]
Sorry for the delay here - this lgtm!
